libbeat/common/kubernetes/watcher.go

```diff
@@ -0,0 +1,263 @@
+	ctx                 context.Context
+	stop                context.CancelFunc
+	bus                 bus.Bus
+	pods                map[string]*Pod      // pod id -> Pod
+	deleted             map[string]time.Time // deleted annotations key -> last access time
+}
+
+// Client for Kubernetes interface
+type Client interface {
+	ListPods(ctx context.Context, namespace string, options ...k8s.Option) (*corev1.PodList, error)
+	WatchPods(ctx context.Context, namespace string, options ...k8s.Option) (*k8s.CoreV1PodWatcher, error)
+}
+
+// NewWatcher initializes the watcher client to provide a local state of
+// pods from the cluster (filtered to the given host)
+func NewWatcher(client Client, syncPeriod, cleanupTimeout time.Duration, host string) Watcher {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &podWatcher{
+		client:              client,
+		cleanupTimeout:      cleanupTimeout,
+		syncPeriod:          syncPeriod,
+		nodeFilter:          k8s.QueryParam("fieldSelector", "spec.nodeName="+host),
+		lastResourceVersion: "0",
+		ctx:                 ctx,
+		stop:                cancel,
+		pods:                make(map[string]*Pod),
+		deleted:             make(map[string]time.Time),
+		bus:                 bus.New("kubernetes"),
+	}
+}
+
+func (p *podWatcher) syncPods() error {
+	logp.Info("kubernetes: %s", "Performing a pod sync")
+	pods, err := p.client.ListPods(
+		p.ctx,
+		"",
+		p.nodeFilter,
+		k8s.ResourceVersion(p.lastResourceVersion))
+
+	if err != nil {
+		return err
+	}
+
+	p.Lock()
+	for _, apiPod := range pods.Items {
+		pod := GetPod(apiPod)
+		p.pods[pod.Metadata.UID] = pod
+	}
+	p.Unlock()
+
+	// Emit all start events (avoid blocking if the bus get's blocked)
+	go func() {
+		for _, pod := range p.pods {
+			p.bus.Publish(bus.Event{
+				"start": true,
+				"pod":   pod,
+			})
+		}
+	}()
+
+	// Store last version
+	p.lastResourceVersion = pods.Metadata.GetResourceVersion()
+
+	logp.Info("kubernetes: %s", "Pod sync done")
+	return nil
+}
+
+// Start watching pods
+func (p *podWatcher) Start() error {
+
+	// Make sure that events don't flow into the annotator before informer is fully set up
+	// Sync initial state:
+	synced := make(chan struct{})
+	go func() {
+		p.syncPods()
+		close(synced)
+	}()
+
+	select {
+	case <-time.After(p.syncPeriod):
+		p.Stop()
+		return errors.New("Timeout while doing initial Kubernetes pods sync")
+	case <-synced:
+		// Watch for new changes
+		go p.watch()
+		go p.cleanupWorker()
+		return nil
+	}
+}
+
+func (p *podWatcher) watch() {
+	for {
+		logp.Info("kubernetes: %s", "Watching API for pod events")
+		watcher, err := p.client.WatchPods(p.ctx, "", p.nodeFilter)
+		if err != nil {
+			//watch pod failures should be logged and gracefully failed over as metadata retrieval
+			//should never stop.
+			logp.Err("kubernetes: Watching API error %v", err)
+			time.Sleep(time.Second)
+			continue
+		}
+
+		for {
+			_, apiPod, err := watcher.Next()
+			if err != nil {
```

Comment:
There is a problem I met before while testing.
This [errors](https://github.com/ericchiang/k8s/issues/68) makes it doing WatchPods forever and never stop. So we might need to check the err type returned by Next.
If it is an ioErr, we should retry watch, but if it is an protocol error, maybe skip?

Reply:
Skip may be risky here? At the end, the break will respawn a new watcher and restart the loop, sounds like a safe bet. If the error is permanent, the watcher creation will sleep for a while.

Reply:
Watcher creation will not sleep, since `p.client.WatchPods` will not return error, the error was returned by `watcher.Next`, and then it break the inner for, and `p.client.WatchPods` immediately since watcher creation will sleep only when `p.client.WatchPods` got error.

Reply:
You are right sorry, WatchPods will sleep after error

Reply:
Also, it needs to Close the watcher before break, or it will be leaked here.

Reply:
Good point, I've pushed the Close call